ml-proto/sexp.py

```diff
@@ -0,0 +1,234 @@
+#!/usr/bin/env python
+
+class SExp(object):
+    def __init__(self, thing, begin=-1, end=-1):
+        assert (begin == -1) == (end == -1)
+        assert not isinstance(thing, SExp)
+        self.thing = thing
+        self.begin = begin
+        self.end = end
+
+    def __repr__(self):
+        return str(self.thing) + ':' + str(self.begin) + ':' + str(self.end)
+
+    def __str__(self):
+        return str(self.thing)
+
+    def has_location(self):
+        return self.begin != -1
+
+    def is_operator(self, name):
+        return isinstance(self.thing, list) and len(self.thing) > 0 and self.thing[0].thing == name
+
+    def num_operands(self):
+        return len(self.thing) - 1
+
+    def get_operand(self, index):
+        return self.thing[index + 1]
+
+    def strip(self):
+        if isinstance(self.thing, list):
+            for item in self.thing:
+                item.strip()
+        self.begin = -1
+        self.end = -1
+
+def isspace(c):
+    return c == ' ' or c == '\t' or c == '\n'
+
+def isdigit(c):
+    return c >= '0' and c <= '9'
+
+def isalnum(c):
+    return isdigit(c) or (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z')
+
+def skip_whitespace_and_comments(s, i):
```

Comment:
This is the only function, in which you are checking if the index exceeds the length of the string. MASM still raises an IndexError, if the index is out-of-bounds in the other functions, but it could be hard to debug mast-files.